python/tests/test_imputation.py

```diff
@@ -0,0 +1,256 @@
+0,3,0.000000,1.000000,0.999900,0.000100,1,1,0.000000,1.000000,2.999800,0.999900
+1,0,1.000000,0.000000,0.999900,0.000100,0,0,0.250000,0.000000,0.249975,0.249975
+1,1,1.000000,0.000000,0.999900,0.000100,1,0,0.250000,0.000000,0.250000,0.000025
+1,2,1.000000,0.000000,0.999900,0.000100,0,0,0.250000,0.000000,0.499975,0.249975
+1,3,1.000000,0.000000,0.999900,0.000100,0,0,0.250000,0.000000,0.749950,0.249975
+2,0,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.249975,0.249975
+2,1,1.000000,0.000000,0.999900,0.000100,0,1,0.250000,0.000000,0.250000,0.000025
+2,2,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.499975,0.249975
+2,3,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.749950,0.249975
+3,0,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.249975,0.249975
+3,1,1.000000,0.000000,0.999900,0.000100,0,1,0.250000,0.000000,0.250000,0.000025
+3,2,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.499975,0.249975
+3,3,1.000000,0.000000,0.999900,0.000100,1,1,0.250000,0.000000,0.749950,0.249975
+"""
+
+_bwd_matrix_text_2 = """
+m,h,probRec,probNoRec,noErrProb,errProb,refAl,queryAl,shift,scale,sum,val
+3,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,1.000000
+3,1,-1,-1,-1,-1,-1,-1,-1,-1,-1,1.000000
+3,2,-1,-1,-1,-1,-1,-1,-1,-1,-1,1.000000
+3,3,-1,-1,-1,-1,-1,-1,-1,-1,-1,1.000000
+2,0,1.000000,0.000000,0.999900,0.000100,1,1,0.000000,0.250000,0.749950,0.250000
+2,1,1.000000,0.000000,0.999900,0.000100,0,1,0.000000,0.250000,0.749950,0.250000
+2,2,1.000000,0.000000,0.999900,0.000100,1,1,0.000000,0.250000,0.749950,0.250000
+2,3,1.000000,0.000000,0.999900,0.000100,1,1,0.000000,0.250000,0.749950,0.250000
+1,0,1.000000,0.000000,0.999900,0.000100,0,1,0.000000,0.250000,0.749950,0.250000
+1,1,1.000000,0.000000,0.999900,0.000100,1,1,0.000000,0.250000,0.749950,0.250000
+1,2,1.000000,0.000000,0.999900,0.000100,0,1,0.000000,0.250000,0.749950,0.250000
+1,3,1.000000,0.000000,0.999900,0.000100,0,1,0.000000,0.250000,0.749950,0.250000
+0,0,1.000000,0.000000,0.999900,0.000100,1,0,0.000000,0.250000,0.749950,0.250000
+0,1,1.000000,0.000000,0.999900,0.000100,0,0,0.000000,0.250000,0.749950,0.250000
+0,2,1.000000,0.000000,0.999900,0.000100,1,0,0.000000,0.250000,0.749950,0.250000
+0,3,1.000000,0.000000,0.999900,0.000100,1,0,0.000000,0.250000,0.749950,0.250000
+"""
+
+
+def convert_to_numpy(matrix_text):
+    """
+    Converts a matrix in text to numpy and returns it.
+    """
+    x = np.loadtxt(io.StringIO(matrix_text), skiprows=1, delimiter=",")
+    for i in np.arange(x.shape[0]):
+        # Check that switch and non-switch probabilities sum to 1
+        assert (x[i, 2] + x[i, 3]) == 1 or x[i, 2] == -1
+        # Check that non-mismatch and mismatch probabilities sum to 1
+        assert (x[i, 4] + x[i, 5]) == 1 or x[i, 4] == -1
+    return x[:, -1].reshape((4, 4))  # size (m, h)
+
+
+def get_forward_backward_matrices():
+    fwd_matrix_1 = convert_to_numpy(_fwd_matrix_text_1)
```

Comment:
I think we can pull in pandas here, to make things a bit easier?

Reply:
We'd add it to the various requirements.txt files. Pulling it in for the tests is fine I think.